From: Rob Evers <revers@redhat.com>
Date: Tue, 5 Jan 2010 15:49:36 -0500
Subject: [scsi] lpfc: fix FC ports offlined during target controller faults
Message-id: 20100105204942.6334.63094.sendpatchset@localhost.localdomain
O-Subject: [RHEL5.4 Z-STREAM PATCH V2] lpfc - Emulex FC ports on RHEL 5.4 GA offlined during target controller faults
Bugzilla: 549906
RH-Acked-by: Pete Zaitcev <zaitcev@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=549906

Fixes problem where FC ports offlined during target controller faults.
This patch was provided by Emulex.

Verified code is in current rhel5.5 lpfc driver

Sanity tested driver by building it on x86-64 rhel5.4 system.  With this patch
applied, the ran a few minutes of data integrity testing using dt.


diff -urpN a/drivers/scsi/lpfc/lpfc_sli.c b/drivers/scsi/lpfc/lpfc_sli.c
--- a/drivers/scsi/lpfc/lpfc_sli.c	2009-11-12 14:37:38.387234000 -0500
+++ b/drivers/scsi/lpfc/lpfc_sli.c	2009-11-12 14:39:18.944732000 -0500
@@ -7746,7 +7746,7 @@ irqreturn_t
 lpfc_sli_sp_intr_handler(int irq, void *dev_id, struct pt_regs *regs)
 {
 	struct lpfc_hba  *phba;
-	uint32_t ha_copy;
+	uint32_t ha_copy, hc_copy;
 	uint32_t work_ha_copy;
 	unsigned long status;
 	unsigned long iflag;
@@ -7804,8 +7804,13 @@ lpfc_sli_sp_intr_handler(int irq, void *
 		}
 
 		/* Clear up only attention source related to slow-path */
+		hc_copy = readl(phba->HCregaddr);
+		writel(hc_copy & ~(HC_MBINT_ENA | HC_R2INT_ENA |
+			HC_LAINT_ENA | HC_ERINT_ENA),
+			phba->HCregaddr);
 		writel((ha_copy & (HA_MBATT | HA_R2_CLR_MSK)),
 			phba->HAregaddr);
+		writel(hc_copy, phba->HCregaddr);
 		readl(phba->HAregaddr); /* flush */
 		spin_unlock_irqrestore(&phba->hbalock, iflag);
 	} else
@@ -8128,6 +8133,7 @@ lpfc_sli_intr_handler(int irq, void *dev
 	struct lpfc_hba  *phba;
 	irqreturn_t sp_irq_rc, fp_irq_rc;
 	unsigned long status1, status2;
+	uint32_t hc_copy;
 
 	/*
 	 * Get the driver's phba structure from the dev_id and
@@ -8165,7 +8171,12 @@ lpfc_sli_intr_handler(int irq, void *dev
 	}
 
 	/* Clear attention sources except link and error attentions */
+	hc_copy = readl(phba->HCregaddr);
+	writel(hc_copy & ~(HC_MBINT_ENA | HC_R0INT_ENA | HC_R1INT_ENA
+		| HC_R2INT_ENA | HC_LAINT_ENA | HC_ERINT_ENA),
+		phba->HCregaddr);
 	writel((phba->ha_copy & ~(HA_LATT | HA_ERATT)), phba->HAregaddr);
+	writel(hc_copy, phba->HCregaddr);
 	readl(phba->HAregaddr); /* flush */
 	spin_unlock(&phba->hbalock);
 
diff -urpN a/drivers/scsi/lpfc/lpfc_version.h b/drivers/scsi/lpfc/lpfc_version.h
--- a/drivers/scsi/lpfc/lpfc_version.h	2009-11-12 14:37:38.395230000 -0500
+++ b/drivers/scsi/lpfc/lpfc_version.h	2009-11-12 14:39:18.950732000 -0500
@@ -18,7 +18,7 @@
  * included with this package.                                     *
  *******************************************************************/
 
-#define LPFC_DRIVER_VERSION "8.2.0.48.2p"
+#define LPFC_DRIVER_VERSION "8.2.0.48.3p"
 
 #define LPFC_DRIVER_NAME		"lpfc"
 #define LPFC_SP_DRIVER_HANDLER_NAME	"lpfc:sp"

